File: python/source/interactive_books/domain/book.py
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum

from interactive_books.domain.errors import BookError, BookErrorCode


class BookStatus(Enum):
    PENDING = "pending"
    INGESTING = "ingesting"
    READY = "ready"
    FAILED = "failed"


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


@dataclass
class Book:
    id: str
    title: str
    status: BookStatus = BookStatus.PENDING
    current_page: int = 0
    embedding_provider: str | None = None
    embedding_dimension: int | None = None
    created_at: datetime = field(default_factory=_utc_now)
    updated_at: datetime = field(default_factory=_utc_now)

    def __post_init__(self) -> None:
        if not self.title.strip():
            raise BookError(BookErrorCode.INVALID_STATE, "Book title cannot be empty")

    def start_ingestion(self) -> None:
        if self.status != BookStatus.PENDING:
            raise BookError(
                BookErrorCode.INVALID_STATE,
                f"Cannot start ingestion from '{self.status.value}' status",
            )
        self.status = BookStatus.INGESTING

    def complete_ingestion(self) -> None:
        if self.status != BookStatus.INGESTING:
            raise BookError(
                BookErrorCode.INVALID_STATE,
                f"Cannot complete ingestion from '{self.status.value}' status",
            )
        self.status = BookStatus.READY

    def fail_ingestion(self) -> None:
        if self.status != BookStatus.INGESTING:
            raise BookError(
                BookErrorCode.INVALID_STATE,
                f"Cannot fail ingestion from '{self.status.value}' status",
            )
        self.status = BookStatus.FAILED

    def reset_to_pending(self) -> None:
        self.status = BookStatus.PENDING

    def set_current_page(self, page: int) -> None:
        if page < 0:
            raise BookError(
                BookErrorCode.INVALID_STATE,
                f"Current page cannot be negative: {page}",
            )
        self.current_page = page

    def switch_embedding_provider(self, provider: str, dimension: int) -> None:
        self.embedding_provider = provider
        self.embedding_dimension = dimension
        self.reset_to_pending()

File: swift/InteractiveBooks/Sources/InteractiveBooksCore/Domain/Entities/Book.swift
import Foundation

public enum BookStatus: String, Sendable, Equatable {
    case pending
    case ingesting
    case ready
    case failed
}

public final class Book: @unchecked Sendable, Equatable {
    public let id: String
    public private(set) var title: String
    public private(set) var status: BookStatus
    public private(set) var currentPage: Int
    public private(set) var embeddingProvider: String?
    public private(set) var embeddingDimension: Int?
    public let createdAt: Date
    public private(set) var updatedAt: Date

    public init(
        id: String,
        title: String,
        status: BookStatus = .pending,
        currentPage: Int = 0,
        embeddingProvider: String? = nil,
        embeddingDimension: Int? = nil,
        createdAt: Date = .now,
        updatedAt: Date = .now,
    ) throws {
        guard !title.trimmingCharacters(in: .whitespaces).isEmpty else {
            throw BookError.invalidState("Book title cannot be empty")
        }
        self.id = id
        self.title = title
        self.status = status
        self.currentPage = currentPage
        self.embeddingProvider = embeddingProvider
        self.embeddingDimension = embeddingDimension
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    public func startIngestion() throws {
        guard status == .pending else {
            throw BookError.invalidState(
                "Cannot start ingestion from '\(status.rawValue)' status",
            )
        }
        status = .ingesting
    }

    public func completeIngestion() throws {
        guard status == .ingesting else {
            throw BookError.invalidState(
                "Cannot complete ingestion from '\(status.rawValue)' status",
            )
        }
        status = .ready
    }

    public func failIngestion() throws {
        guard status == .ingesting else {
            throw BookError.invalidState(
                "Cannot fail ingestion from '\(status.rawValue)' status",
            )
        }
        status = .failed
    }

    public func resetToPending() {
        status = .pending
    }

    public func setCurrentPage(_ page: Int) throws {
        guard page >= 0 else {
            throw BookError.invalidState(
                "Current page cannot be negative: \(page)",
            )
        }
        currentPage = page
    }

    public func switchEmbeddingProvider(provider: String, dimension: Int) {
        embeddingProvider = provider
        embeddingDimension = dimension
        resetToPending()
    }

    package init(
        fromRow id: String,
        title: String,
        status: BookStatus,
        currentPage: Int,
        embeddingProvider: String?,
        embeddingDimension: Int?,
        createdAt: Date,
        updatedAt: Date,
    ) {
        self.id = id
        self.title = title
        self.status = status
        self.currentPage = currentPage
        self.embeddingProvider = embeddingProvider
        self.embeddingDimension = embeddingDimension
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    public static func == (lhs: Book, rhs: Book) -> Bool {
        lhs.id == rhs.id
    }
}
